### PR TITLE
support removing `class-string`s from `GenericClassStringType`

### PR DESCRIPTION
**Pull request by staabm**
refs https://github.com/phpstan/phpstan/issues/7698#issuecomment-1202345351

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Please also add a regression test for the match expression, there's also some example with $obj::class I think.

**Comment by Dgame** [comment]
Here is a smaller example: https://phpstan.org/r/e4476215-99ea-434f-adae-3564ef93a5a5

If this works I really have to take a look how the semantics work, I'm not sure I understand the NeverType there.

**Comment by Dgame** [comment]
Not sure if this is related in this case, but  have another test case there: https://github.com/phpstan/phpstan/issues/7721

**Comment by staabm** [comment]
just added the match-test. its still failling though.

I think Dgame is right, that there is a missing piece in `MatchExpressionRule`. I need to invest more time, to get an idea how this match-arm analysis is working... 
it seems `MatchExpressionRule` is successfully removing all match-arm expressions from the match condition, but instead of a `NeverType` it leads to a empty `ClassStringType` atm

**Comment by staabm** [comment]
ok, after sleeping a night over it, I know understood the basic building blocks of the arm analysis.

we more or less rewrite the AST in 
https://github.com/phpstan/phpstan-src/blob/1537424191a4cf27577dd3cfc07d162b09c20a12/src/Rules/Comparison/MatchExpressionRule.php#L54-L57

which adds similar IF-identical expressions we already have in https://github.com/phpstan/phpstan-src/pull/1590/files#diff-9ca94a9a28f6d393c869ef34bde57e4bb0d53dcc9a8cb5fc865f074ef0df5348R30-R32

with this knowledge I was able to fix the test with commit [21e5bcf](https://github.com/phpstan/phpstan-src/pull/1590/commits/21e5bcf058cd5963ee5f0db5e9bfbb63d5a1f436).

----

still, I think the match rule contains a bug which makes it only work when the match condition is a variable.
as soon as you put directly a expression into the match, it does not work atm:

works:
```php
	$t = new Test(new A());
	$class = $t->value::class;
	echo match ($class) {
		A::class => 'A',
		B::class => 'B'
	};
```

doesn't work:
```php
	$t = new Test(new A());
	echo match ($t->value::class) {
		A::class => 'A',
		B::class => 'B'
	};
```

I am still figuring out the details.

**Comment by staabm** [comment]
I put my finding into a new issue, as I think this one can be merged as is?

https://github.com/phpstan/phpstan/issues/7746

**Comment by Dgame** [comment]
Thanks @staabm! That a good step forward. 🙂 Could we call `tryRemove` in the MatchExpressionRule in case the match-condition is an expression?

**Comment by staabm** [comment]
issue https://github.com/phpstan/phpstan/issues/7746 is not related to the MatchRule. its a more general problem, see e.g. https://phpstan.org/r/be1b13ce-b8d9-445e-b6ed-dc9e87f2879b

**Comment by Dgame** [comment]
@staabm Here is a more reduced example which does not work: https://phpstan.org/r/0a06314d-d25e-4435-8aa1-d60bb41b8fda
It does not seems to work if I store the class-name in a variable.

**Comment by staabm** [comment]
phpstan.org/try does not yet contain the patch, as the PR is not yet merged.
taking your example and running it against phpstan with the PR applied works for me as expected locally (no errors)

File: src/Type/Generic/GenericClassStringType.php [hunk 1]
```diff
@@ -172,4 +173,24 @@ public static function __set_state(array $properties): Type
 		return new self($properties['type']);
 	}
 
+	public function tryRemove(Type $typeToRemove): ?Type
+	{
+		if ($typeToRemove instanceof ConstantStringType && $typeToRemove->isClassString()) {
+			$generic = $this->getGenericType();
+
+			if ($generic instanceof ObjectType) {
```
Comment (ondrejmirtes) [inline src/Type/Generic/GenericClassStringType.php]: TypeWithClassName pls

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Generic/GenericClassStringType.php (hunk 1 above)
Comment (staabm) [reply src/Type/Generic/GenericClassStringType.php]: I tried that while PR creation and it made some tests fail.

Therefore I went back to ObjectType

Comment (ondrejmirtes) [reply src/Type/Generic/GenericClassStringType.php]: That's a red flag, we should investigate that

Comment (staabm) [reply src/Type/Generic/GenericClassStringType.php]: ok - seems to work now. maybe it was some intermediate step which made it fail…

**Comment by ondrejmirtes** [comment]
Thank you.

**Comment by Dgame** [comment]
Thanks a lot @staabm !

**Pull request merged**